Feature/parsing counting

kuchichan opened the pull request:
Ready to review

m-kop committed:
adding input module

* initial commit of input module
* add test harness for input module
* add inputTestDirecotry/ with example input files

m-kop committed:
add basic test of input path

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,17 @@
+#include <string>
+#include <vector>
+
+class FrameParser {
```
In my opinion FrameParser shoud have static function to parse frameString and return vector of frames.

kuchichan replied:
We will see.

kszytko replied:
I was wrong,  static function is not good idea. We should operate here on normal object.

kuchichan replied:
I will leave It as it is.

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,17 @@
+public:
+    FrameParser(const std::string& gameString);  // thinking about explicit...
+
+    std::string getGameString() const { return gameString_; }
```
redundant

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,17 @@
+    std::string getGameString() const { return gameString_; }
+    std::vector<size_t> getFrames() const { return frames_; }
+
+    void parse();
```
can be turn into  static function
```suggestion
    static std::vector<size_t> parse(const std::string& gamePoints){
        //parsingString into frames vector
        return frames;
    }
```

kuchichan replied:
This is vip, we will see how it goes.

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,17 @@
+    FrameParser(const std::string& gameString);  // thinking about explicit...
+
+    std::string getGameString() const { return gameString_; }
+    std::vector<size_t> getFrames() const { return frames_; }
```
redundant

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,17 @@
+
+private:
+    std::string gameString_;
+    std::vector<size_t> frames_;
```
redundant

kszytko commented on bowling/inc/game.hpp:
```diff
@@ -2,10 +2,10 @@
-constexpr int MAX_FRAMES{ 10 };
-constexpr int MAX_ROLLS{ 21};
-constexpr int MAX_FRAME_SCORE{ MAX_PINS };
+constexpr int MAX_PINS{10};
+constexpr int MAX_FRAMES{10};
+constexpr int MAX_ROLLS{21};
+constexpr int MAX_FRAME_SCORE{MAX_PINS};
```
do not modify other class if you work only on frameParser.  Make it different PR

kuchichan replied:
Agreed, done.

kszytko commented on bowling/src/main.cpp:
```diff
@@ -1,5 +1,3 @@
-int main() {
-
-}
+int main() {}
```
same up

kszytko commented on bowling/src/game.cpp:
```diff
@@ -9,12 +9,10 @@ int Game::score() {
         if (isStrike(firstFrameThrow)) {
-            score_ += MAX_FRAME_SCORE +
-                      nextTwoBallsForStrike(firstFrameThrow);
+            score_ += MAX_FRAME_SCORE + nextTwoBallsForStrike(firstFrameThrow);
```
same up

kszytko reviewed (changes requested):
Some of ideas to make. Modifiyng formating in other classes should go into other PR

kuchichan committed:
remove clang format changes

kuchichan committed:
remove clang format changes

kuchichan committed:
remove clang format changes

kuchichan committed:
remove clang format changes

m-kop committed:
fix for build in CI

CI requires filesystem library to be used as "experimanetal"
(old GCC requiremenets)

m-kop committed:
add player and lane classes

m-kop committed:
add reading files names

m-kop committed:
add Lanes to Input

m-kop committed:
refactor Input class

m-kop committed:
add reading players

kuchichan committed:
Working, needed tests + refactor

kuchichan committed:
Working, needed tests + refactor

kuchichan committed:
Working, needed tests + refactor

kuchichan committed:
Working, needed tests + refactor

kuchichan committed:
algo + rename

kuchichan committed:
algo + rename

kuchichan committed:
algo + rename

kuchichan committed:
algo + rename

m-kop committed:
refactor - parsing players not in the scope of this work

m-kop committed:
refator to use share_ptr for collection of Lanes

m-kop committed:
add players as strings + tests

m-kop committed:
add lanes_ initialization

m-kop committed:
moving init to Input constructor

m-kop committed:
add init

m-kop committed:
add init

m-kop committed:
github debug

m-kop committed:
add ASSERT isValid() in test

m-kop committed:
modify path to test directory

m-kop committed:
Revert "github debug"

This reverts commit 81948f96aea2ae96a3d023df4a79383284f24a70.

m-kop committed:
update after review

m-kop committed:
refactored tests

m-kop committed:
apply review comments

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,22 @@
+#include <string>
+#include <vector>
+
+constexpr char DELIMITER = '|';
+constexpr char STRIKE = 'X';  // Maybe map like - strike - 10, spare miss, 0, but additional logic with spare sign...
+constexpr char SPARE_SIGN = '/';
+constexpr char MISS = '-';
```
move it into class,   only class need it.    
Minus is that constexpr doesn't work in classes, instead use normal const.

kszytko replied:
And consider to use or not `_SIGN` in every const

kuchichan replied:
Done.

kszytko commented on bowling/inc/frameParser.hpp:
```diff
@@ -0,0 +1,22 @@
+};
+
+class Tokenizer {
+public:
```
i dont think this class in needed

kuchichan replied:
I think it is needed, since its easier to test and we are not implementing antipattern "iceberg" class

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+		 });
+  
+  return std::pair(tokenized.first, parsed);
+}
```
to long function,  split to smallest private methods

kuchichan replied:
Done.

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+  std::transform(tokenized.second.cbegin(),
+		 tokenized.second.cend(),
+		 back_inserter(parsed),
+		 [&last_parsed_char] (char el) {
+		   if (el == 'X') {
+		     last_parsed_char = 10;
+		     return last_parsed_char;
+		   } else if (el == '/') {
+		     last_parsed_char = 10 - last_parsed_char;
+		     return last_parsed_char;
+		   } else if (el == '-') {
+		     last_parsed_char = 0;
+		     return last_parsed_char;
+		   } else {
+		     last_parsed_char = el - '0';
+		     return last_parsed_char;
+		   }
+		 });
```
too big lambda function

kszytko replied:
Extract it into method

kuchichan replied:
Done.

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+  std::vector<size_t> parsed {};
+  size_t last_parsed_char{};
+  auto tokenized = Tokenizer::splitToFramesAlt(gamePoints);
+  std::transform(tokenized.second.cbegin(),
+		 tokenized.second.cend(),
```
bad idea to modify item of pair like this. very bad

kuchichan replied:
Could you elaborate, why? Maybe some source, materials. Im thinking about leaving pairs at all in favour of just simple vector where first element will be the name - im just little bit worried for example about strings without names.

kszytko replied:
Becouse as reader i dont know what this pair holds,  better way is to use separate variables for name and array

kuchichan replied:
Done.

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+parsedFrame FrameParser::parse(const std::string& gamePoints) {
+  std::vector<size_t> parsed {};
+  size_t last_parsed_char{};
+  auto tokenized = Tokenizer::splitToFramesAlt(gamePoints);
```
hmm,  maybe better name should be  pointList ? parsedList?   what exaclty splitToFramesAlt does?

kszytko replied:
split it into two methods something like: 
```suggestion
  auto playerName = readPlayerName(gameLine);
  auto gameSequence = readSequence(gameLine);
```
you dont need to use pairs

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+		 back_inserter(parsed),
+		 [&last_parsed_char] (char el) {
+		   if (el == 'X') {
+		     last_parsed_char = 10;
```
we use camelCase variable convention

kszytko replied:
what exacly last_parsed_char hold?   why not char?  name does not tell us anything

kuchichan replied:
last_parser_char holds last calculated "shot " and its use to calculate exact value from "/" -> that is the way that is used in game.cpp

kuchichan replied:
Made somewhat easier to read.

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+      prev = next + 1;
+    }
+    
+    std::vector<char> result = {
+        std::next(gamePoints.begin(), prev),
+        gamePoints.end()};
```
you did something, but it's not readable totally

kuchichan replied:
Why? I Just use the fact, that vector can take iterator ranges, and this is efficient way to switch from string to vector of chars.

kszytko replied:
String is a vector of char itself

kuchichan replied:
That's right. Done.

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+        std::next(gamePoints.begin(), prev),
+        gamePoints.end()};
+    
+    result.erase(std::remove_if(result.begin(), result.end(),
+				[] (char el) {return el == '|'; }),
+		 result.end());
```
```suggestion
    auto separator = '|';  //or make it const
    result.erase(std::remove(result.begin(), result.end(),  separator ),
		 result.end());
```

kuchichan replied:
I will fix all these magic chars (look constexpr -> const in class)

kuchichan replied:
Done.

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+}
+
+tokenizedFrame Tokenizer::splitToFramesAlt(const std::string& gamePoints) {
+    size_t prev = 0;
+    size_t next = 0;
+    std::string name = "";
+    
+    next = gamePoints.find(":", prev);
+    
+    if (next != std::string::npos) {
+      name = gamePoints.substr(prev, next - prev);
+      prev = next + 1;
+    }
```
Separate into 2 methods, one for readName, second for matchSequence

kszytko replied:
Question!  What type of data will you get from folderReader class?   lineString without parsed name?

kuchichan replied:
I do not know ;p I assumed, that i have to parse all. Due to your proposals /advice its easy to turn this part off :)

kszytko commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,54 @@
+parsedFrame FrameParser::parse(const std::string& gamePoints) {
+  std::vector<size_t> parsed {};
+  size_t last_parsed_char{};
+  auto tokenized = Tokenizer::splitToFramesAlt(gamePoints);
```
Separate into 2 methods, f.e. one for readName, second for readSequence:
```suggestion
  auto name= readName(line);
  auto sequence = readSequence(line);
```

kuchichan replied:
Done.

kszytko reviewed (changes requested):
Some changes,  please write tests.

niemarcin committed:
argumentParser unit tests refactored

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
tests + major refactor

kuchichan committed:
additional test

kuchichan committed:
additional test

kuchichan committed:
fix

kuchichan committed:
fix

niemarcin committed:
code formatting

m-kop committed:
Merge remote-tracking branch 'origin/bowling' into handling_input_files

m-kop committed:
add 2nd review comments

m-kop committed:
add RC

m-kop committed:
add exception throwing

kuchichan committed:
Small after rebase

kuchichan committed:
Small after rebase

kuchichan committed:
Delete non needed file

kuchichan committed:
Delete non needed file

kuchichan committed:
CmakeLists

kuchichan committed:
Small tweaking in files

kuchichan committed:
Small tweaking in files

kuchichan committed:
Trailing whitespace

kuchichan committed:
Trailing whitespace

niemarcin commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,40 @@
+        lastParsedToken = token - '0';
+        return lastParsedToken;
+    }
+}
```
1. You don't need to `return lastParsedToken;` in every condition. Just do it at the end of the method.
2. Do you think switch-case instead of if-else could be better?

kuchichan replied:
Done.

niemarcin commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,40 @@
+const char FrameParser::SpareSign = '/';
+
+
+size_t FrameParser::parseToken(const char token, size_t& lastParsedToken) {
```
It seems inconsistent that `parseToken` method is used to parse a token which is `const char` and in the same line we have `lastParseToken` which type is `size_t&`. Both are tokens (current and previous) but have different types. 

Maybe the `lastParsedToken` name could be changed to something different? `lastTokenValue` or `previousTokenValue` or something like that.

niemarcin replied:
Another thing - I don't know if it it's a good practise to return the same value by `return` statement from function and at the same time through a reference to parameter. I can understand why it's used like this by looking at  `parse()` method, but I'm just wondering if it's good or bad to use it like this.

kuchichan replied:
To be honest, I have also feeling that it is not the best way to do that. I have tried with lambda function and capture list, but this was also derided in previous review. I do not really have idea what to do. I will need this last parsed element anyway. 

To the first comment - please note that is actually "calculated" value (integer/size_t) of token 10, 5 ,3 or so, not token itself '/' '3'.

kuchichan replied:
I will leave as it is (maybe there is obvoius way for doing this - there kind of problems are probably common).

niemarcin replied:
> To the first comment - please note that is actually "calculated" value (integer/size_t) of token 10, 5 ,3 or so, not token itself '/' '3'.

Then maybe `lastRollValue` or just `lastValue` ?

niemarcin replied:
And related to other stuff:

I've tried this:
```cpp
void FrameParser::parseToken(const char token, size_t& lastParsedToken) {
    switch (token) {
    case FrameParser::StrikeSign:
        lastParsedToken = 10;
        break;
    case FrameParser::SpareSign:
        lastParsedToken = 10 - lastParsedToken;
        break;
    case FrameParser::MissSign:
        lastParsedToken = 0;
        break;
    default:
        lastParsedToken = static_cast<size_t>(token - '0');
        break;
    }
}
```
As you can see I've changed return type to `void`.

And in `FrameParser::parse()` method, instead of `std::transform` I've called this:
```cpp
    std::for_each(sequence.cbegin(), sequence.cend(),
                  [&lastParsedToken, &parsed](const auto& el){
                      FrameParser::parseToken(el, lastParsedToken);
                      parsed.push_back(lastParsedToken);
                  });
```

Just my suggestion, not the ultimate solution.  ;) But tests are still fine.

kuchichan replied:
Done.

niemarcin commented on bowling/src/frameParser.cpp:
```diff
@@ -0,0 +1,40 @@
+        lastParsedToken = 0;
+        return lastParsedToken;
+    } else {
+        lastParsedToken = token - '0';
```
Explicite type cast might be better.

kuchichan replied:
Could you be more specific? Like static_cast<size_t> ? This will not produce needed output.

niemarcin replied:
Why not? I've just checked `lastParsedToken = static_cast<size_t>(token - '0');` - it compiles and tests still pass.

I'm not sure if it's better to use implicit or explicit casting, but based on experience with other languages I believe it's better to avoid implicit type conversion.

kuchichan replied:
Done.

niemarcin commented on bowling/src/tokenizer.cpp:
```diff
@@ -0,0 +1,23 @@
+const char Tokenizer::FrameDelimiterSign = '|';
+const char Tokenizer::NameDelimiterSign = ':';
+
+std::string Tokenizer::readSequence(const std::string& line, size_t& position) {
+    std::string result = {std::next(line.begin() + position), line.end()};
+
+    result.erase(std::remove(result.begin(), result.end(), FrameDelimiterSign), result.end());
+
+    return result;
+}
+
+std::string Tokenizer::readName(const std::string& line, size_t& position) {
+    size_t next = line.find(NameDelimiterSign);
+    if (next != std::string::npos) {
+        position = next;
+        return line.substr(0, next);
+    } else {
+        return "";
+    }
+}
```
It took me a while to understand how these functions work. 

I don't know if it's a good idea to make it like this. Now `readSequence()` and `readName()` methods are tightly coupled by the `position` parameters and the way how they are called in FrameParser class. Wouldn't be better to just remove `position` from a parameters and once again use `line.find(NameDelimiterSign)` ? I know that it's less efficient but readability and **independence** (no coupling between) of these methods will be better.

So it would be something like this:
```cpp
std::string Tokenizer::readSequence(const std::string& line) {
    size_t position = line.find(NameDelimiterSign);
    if (position == std::string::npos) {
        return "";
    }
    std::string result = {std::next(line.begin() + position), line.end()};
    result.erase(std::remove(result.begin(), result.end(), FrameDelimiterSign), result.end());
    return result;
}

std::string Tokenizer::readName(const std::string& line) {
    size_t position = line.find(NameDelimiterSign);
    if (position != std::string::npos) {
        return line.substr(0, position);
    } else {
        return "";
    }
}
```

And then in `FrameParser::parse`  the `size_t currentLinePosition{};` is not needed.

And of course tokenizer-ut will change
```cpp
TEST_P(GetNameSuite, TestReadNameReturnsPartOfTheStringToColonSign) {
    auto [param, expected] = GetParam();
    ASSERT_EQ(Tokenizer::readName(param), expected);
}

TEST_P(GetSeqSuite, TestReadSequenceReturnsStringWithoutNameAndDelimiter) {
    auto [param, expected] = GetParam();
    ASSERT_EQ(Tokenizer::readSequence(param), expected);
}
```

Or maybe we can find the `NameDelimiterSign` position once and store it somewhere for future use? But finding this delimiter should **not** be done in `readName` method for later use in `readSequence`. Maybe additional `findNameDelimiter()` method should be created?

kuchichan replied:
Done.

niemarcin reviewed (changes requested):
I don't get one thing in general - why both `FrameParser `and `Tokenizer `classes are fully `static`?

kuchichan committed:
Review pt1

kuchichan committed:
Review pt1

kuchichan committed:
Review pt2

kuchichan committed:
Review pt2

niemarcin reviewed (approved):
Great work!

niemarcin committed:
Merge pull request #17 from niemarcin/argumentParserUT

argumentParser unit tests refactored

kszytko reviewed (changes requested):
Parser not working with empty shoots:
```diff
- "1-||X "-> {1,0,10,0},  is
+ "1-||X "-> {1,0,0,0,10,0},   should be
```

kuchichan commented:
> Parser not working with empty shoots:
> 
> ```diff
> - "1-||X "-> {1,0,10,0},  is
> + "1-||X "-> {1,0,0,0,10,0},   should be
> ```

Except that is completely wrong input.

kszytko commented:
> > Parser not working with empty shoots:
> > ```diff
> > - "1-||X "-> {1,0,10,0},  is
> > + "1-||X "-> {1,0,0,0,10,0},   should be
> > ```
> 
> Except that is completely wrong input.

Ok, my mistake.

m-kop committed:
used namespace alias

m-kop committed:
add index check for players_

m-kop committed:
formating for tests

m-kop committed:
formatting

m-kop committed:
more formatting

m-kop committed:
Merge pull request #13 from kszytko/handling_input_files

Handling input files

kuchichan committed:
Merge branch 'feature/parsing-counting' of https://github.com/kuchichan/testing into feature/parsing-counting

kuchichan committed:
Fixing CMakeLists.txt
